src/app/checklist/app-updates/app-updates.service.ts

```diff
@@ -0,0 +1,53 @@
+import { Injectable } from '@angular/core';
+import { MatSnackBar } from '@angular/material';
+import { SwUpdate } from '@angular/service-worker';
+import {
+  fromEvent as observableFromEvent,
+  interval,
+  merge as observableMerge,
+  Observable,
+  of as observableOf
+} from 'rxjs';
+import { map } from 'rxjs/operators';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class AppUpdatesService {
+  constructor(public snackbar: MatSnackBar, private serviceWorkerUpdates: SwUpdate) {}
+
+  checkUpdatesAuto() {
+    this.serviceWorkerUpdates.activated.subscribe(() => {
+      this.snackbar.open('App updated', 'OK', { duration: 2000 });
+    });
+
+    this.serviceWorkerUpdates.available.subscribe(() => {
+      const updateSnackbar = this.snackbar.open('Update available!', 'Update Now', {
+        duration: 2000
+      });
+
+      updateSnackbar.onAction().subscribe(() => {
+        updateSnackbar.dismiss();
+        window.location.reload();
+      });
+    });
+  }
+
+  checkUpdatesManually() {
```

Comment:
I am not so sure about this method and I can see a few issues here.

One problem I see is that for every call of this function, we create an `interval` and it's not being cleaned up until the application is closed. This means if I call this method 3 times then we set up 3 intervals which keep checking for updates. This procedure only needs to be executed once in the background. So I am thinking if it would make more sense to return the interval and let the consumer subscribe. This way the consumer is also responsible for cleaning up the subscription.

Another idea would be to have small conditional that checks if an interval is already running. If yes, then do nothing, if no, then start the interval.

Reply:
Maybe we don't even need this interval logic? Because the `SwUpdate` already does that no? I was thinking of a method to force-update the site and reload the SW including all assets. Curious to hear your thoughts @tayambamwanza.

Reply:
This is straight from the docs

```
import { ApplicationRef, Injectable } from '@angular/core';
import { SwUpdate } from '@angular/service-worker';
import { concat, interval } from 'rxjs';
import { first } from 'rxjs/operators';

@Injectable()
export class CheckForUpdateService {

  constructor(appRef: ApplicationRef, updates: SwUpdate) {
    // Allow the app to stabilize first, before starting polling for updates with `interval()`.
    const appIsStable$ = appRef.isStable.pipe(first(isStable => isStable === true));
    const everySixHours$ = interval(6 * 60 * 60 * 1000);
    const everySixHoursOnceAppIsStable$ = concat(appIsStable$, everySixHours$);

    everySixHoursOnceAppIsStable$.subscribe(() => updates.checkForUpdate());
  }
}
```

Reply:
Off the top of my head: I'm thinking that we only need to keep track of the **isStable** variable, if the app isn't stable yet the button could be hidden or show some sort of indicator on the buttons text like "_initialising_" once the **isStable** variable becomes **true** then the button will be shown or the text will change to say "_check for updates_" 

When a user clicks on the enabled button, a toast will show saying "checking for updates", the problem is that the checkForUpdate() method doesn't seem to indicate whether an update is found or not:

> This method returns a Promise which indicates that the update check has completed successfully, though it does not indicate whether an update was discovered as a result of the check. Even if one is found, the service worker must still successfully download the changed files, which can fail. If successful, the available event will indicate availability of a new version of the app.

thoughts?

Reply:
The interval being recreated here, what @d3lm is referring to, is correct I believe. The example in the angular docs is executed in the constructor, and thus only once. By exposing it in this function here, a consumer can call it multiple times, triggering multiple (unmanaged) subscriptions. 
I think it's best to keep a reference to the observable, add the `shareReplay(1)` operator, and return the same reference on every subsequent call to the method.

Reply:
Regarding the button. Do we need to show a button to check for updates? I was hoping this would be an automatic process with a button shown when there is actually an update.